hack/bin/Sbom.hs

```diff
@@ -0,0 +1,349 @@
+import System.Process
+
+data License = MkLicense
+  { id :: Maybe Text,
+    name :: Maybe Text
+  }
+  deriving stock (Eq, Ord, Show, Generic)
+  deriving anyclass (FromJSON, ToJSON)
+
+sadSbomMeta :: Text -> Text -> [Text] -> SBomMeta Identity
+sadSbomMeta drvPath outPath directDeps =
+  MkSBomMeta
+    { drvPath = drvPath,
+      outPath = Identity outPath,
+      directDeps = Identity directDeps,
+      description = Nothing,
+      homepage = Nothing,
+      licenseSpdxId = [],
+      name = Nothing,
+      typ = Nothing,
+      urls = [],
+      version = Nothing
+    }
+
+data SBomMeta f = MkSBomMeta
+  { drvPath :: Text,
+    description :: Maybe Text,
+    homepage :: Maybe Text,
+    licenseSpdxId :: [Maybe License],
+    name :: Maybe Text,
+    typ :: Maybe Text,
+    urls :: [Maybe Text],
+    version :: Maybe Text,
+    outPath :: f Text,
+    directDeps :: f [Text]
+  }
+
+deriving stock instance (Eq (f [Text]), Eq (f Text)) => Eq (SBomMeta f)
+
+deriving stock instance (Ord (f [Text]), Ord (f Text)) => Ord (SBomMeta f)
+
+deriving stock instance (Show (f [Text]), Show (f Text)) => Show (SBomMeta f)
+
+type Meta = SBomMeta Proxy
+
+instance FromJSON Meta where
+  parseJSON (Object val) =
+    MkSBomMeta
+      <$> val
+      .: "drvPath"
+      <*> val
+      .: "description"
+      <*> val
+      .: "homepage"
+      <*> val
+      .: "licenseSpdxId"
+      <*> val
+      .: "name"
+      <*> val
+      .: "type"
+      <*> val
+      .: "urls"
+      <*> val
+      .: "version"
+      <*> pure Proxy
+      <*> pure Proxy
+  parseJSON invalid = typeMismatch "Object" invalid
+
+type SBom = Map Text (SBomMeta Identity)
+
+type MetaDB = Map Text (SBomMeta Proxy)
+
+type ClosureInfo = Tree ByteString
+
+type PathInfo = [(Text, (Text, [Text]))]
+
+data Visit a = Seen a | Unseen a
+  deriving stock (Eq, Ord, Show)
+
+data SerializeSBom = MkSerializeSBom
+  { -- | the version of the SBom; this is version of the old SBom + 1
+    sbom'version :: Natural,
+    -- | name of the component the SBom is generated for
+    sbom'component :: Text,
+    -- | the creator of the component the SBom is generated for
+    sbom'manufacture :: Text,
+    -- | the supplier (manufacturer or repackager or distributor)
+    sbom'supplier :: Maybe Text,
+    -- | (spdxids of) licenses of the product
+    sbom'licenses :: [Text]
+  }
+
+defaultSerializeSBom :: SerializeSBom
+defaultSerializeSBom =
+  MkSerializeSBom
+    { sbom'version = 1,
+      sbom'component = "wire-server",
+      sbom'manufacture = "wire",
+      sbom'supplier = Nothing,
+      sbom'licenses = ["AGPL-3.0-or-later"]
+    }
+
+-- FUTUREWORK(mangoiv): we can also have
+--
+-- - qualifiers: extra qualifying data for a package such as an OS, architecture, a distro, etc. Optional and type-specific.
+-- - subpath: extra subpath within a package, relative to the package root. Optional.
+-- - use heuristics based approach to finding original repositories for packages, e.g. pkg:hackage....
```

Comment:
@MangoIV would supporting more "original repositories" essentially mean code like this?

````...
      | any (maybe False (T.isInfixOf "hackage.haskell.org")) meta.urls = "hackage"
      | any (maybe False (T.isInfixOf "crates.io")) meta.urls = "cargo"
````

Reply:
yeah I mean if we expect to have the urls not contain any cross references to these domains even though they don't originate for that, that's correct, yes.

Reply:
the issue is that this approach is horrendously slow, I will use something like AhoCorasick if we want to match on more repos.